naps/cores/cmv12k/pixel_remapper.py

```diff
@@ -4,56 +4,334 @@
+
+        self.input_channels = [
+            input.payload[n_bits * i : n_bits * (i + 1)] for i in range(self.n_channels)
+        ]
+
+        self.output_channels = [
+            self.output.payload[n_bits * i : n_bits * (i + 1)] for i in range(self.n_channels // 2)
+        ]
+
+        self.row_width = (
+            4096 * 2
+        )  # we act as if we just have a single row of twice the length of a single row to accomodate the two-sided readout
+
+        self.pixels_per_row_per_channel = self.row_width // self.n_channels
 
-    Generally the CMV12000 sends its data out in bursts.
-    The output format of the CMV12000 (and therefore the input for this core) depends on two factors:
-    the number of lanes used and the presence of subsampling  / binning. The burst length is described by
-    line_length / n_lanes_per_side so for the non line_lengh of 4096 and 16 lanes per side
-    (the default beta configuration) this equals to a burst_length of 256.
-    One Lane always sends adjacent pixels of bust_length out.
+        # fixed sensor parameter
+        self.burst_size = 128
+        self.bursts_per_channel = self.pixels_per_row_per_channel // self.burst_size
 
-    """
-    def __init__(self, top_lanes: List[Signal], bottom_lanes: List[Signal], control_lane: Signal):
-        self.lanes = top_lanes
-        self.bottom_lanes = bottom_lanes
-        self.control_lane = control_lane
-        assert len(top_lanes) == len(bottom_lanes)
+        self.n_mem = self.n_channels // 4
 
-        self.lines_to_buffer = 64 // (len(top_lanes) + len(bottom_lanes))
-        self.n_bits = len(top_lanes[0])
-        self.n_lanes_per_side = len(top_lanes)
+        # each channel writes channel_block_size = row_width / 2 / self.n_mem / self.n_channel words into each BRAM. So we give each a continuous region of that in each BRAM
+        # ordered by the channels, so we can just read all BRAMs from the same address for readout
+        self.channel_block_size = self.row_width // 2 // self.n_mem // self.n_channels
 
-        self.output = ImageStream(self.n_bits)
+        # be able to store two full rows
+        self.buffer_depth = self.channel_block_size * self.n_channels
+        self.mem_depth = self.buffer_depth * 2
+        print("pixel_remapper: mem_depth:", self.mem_depth)
+        self.addr_width = bits_for(self.mem_depth - 1)
 
-    def elaborate(self, platform):
+        self.output_words_per_row = self.row_width // self.n_channels
+
+    def elaborate(self, _):
         m = Module()
 
-        cw = ControlChannelWord(self.control_lane)
+        # all memories have the same value for these
+        rd_enable = Signal()
+        rd_address = Signal(self.addr_width)
+        wr_enable = Signal()
+
+        # write address, write data and rea data is independent for each memory.
+        wr_address = [Signal(self.addr_width, name=f"wr_address_{i:02}") for i in range(self.n_mem)]
```

Comment:
bikeshed: maybe we should rather store the ports here?

Reply:
Sure, don't have a strong feeling either way.